### ISSUE TITLE
Build is unstable

### ISSUE THREAD

**Issue by TommyJones** [TommyJones/tidylda#23]
R CMD check fails randomly across platforms. Usually the failure is due to negative probabilities in the gibbs sampler. Current build passes tests when testing but not running full CMD check. But then the issue crops up in CMD check when running examples or tests. Same code, same data. 

Need a build that  
1. Is stable
2. Uses thread safe containers

**Comment by TommyJones** [comment]
I think the instability is just a continuation of the negative probabilities problem.

Managed to trace it back to the iterator for documents, d. If I initialize d = 0 in the main loop over documents, somehow Cd, Cv, Ck, Zd, and docs get out of sync. When d isn't initialized at 0, I'm not sure why there isn't a segfault. But there isn't, however the model calculates total junk. That junk then creates situations that cause CMD check to fail when checking examples or running unit tests.

Bottom line, I think I need to refactor the code entirely.

**Comment by TommyJones** [comment]
dba358dadf1d27d21532d16e4e66fcbab8e545bd has it (probably) fixed.

Issue was pass by reference vs. pass by value. I was passing by value when I thought I was passing by reference. 

Keeping the issue open to do more checks and optimizations/cleanup of code.

**Comment by TommyJones** [comment]
As of b4a0fd8ef91dae6abafdb4d6b4056b2f5b2353d5 I'm willing to say the build is stable. However the Gibbs sampler is about 3X slower than the one I have in {textmineR}. Opening another issue to address that. :-|
